pkg/operator/events/recorder_upstream.go

```diff
@@ -27,20 +30,58 @@ type upstreamRecorder struct {
 	broadcaster       record.EventBroadcaster
 	eventRecorder     record.EventRecorder
 	involvedObjectRef *corev1.ObjectReference
+
+	// shuttingDown indicates that the broadcaster for this recorder is being shut down
+	shuttingDown  bool
+	shutdownMutex sync.RWMutex
+
+	// fallbackRecorder is used when the kube recorder is shutting down
+	// in that case we create the events directly.
+	fallbackRecorder Recorder
+}
+
+var eventsCounterMetric = metrics.NewCounterVec(&metrics.CounterOpts{
+	Subsystem:      "event_recorder",
+	Name:           "total_events_count",
+	Help:           "Total count of events processed by this event recorder per involved object",
+	StabilityLevel: metrics.ALPHA,
+}, []string{"namespace", "name"})
+
+func init() {
+	(&sync.Once{}).Do(func() {
+		legacyregistry.MustRegister(eventsCounterMetric)
+	})
 }
 
 func (r *upstreamRecorder) ForComponent(componentName string) Recorder {
-	newRecorderForComponent := *r
-	broadcaster := record.NewBroadcaster()
+	newRecorderForComponent := upstreamRecorder{
+		client:            r.client,
+		fallbackRecorder:  r.fallbackRecorder.WithComponentSuffix(componentName),
+		involvedObjectRef: r.involvedObjectRef,
+		shuttingDown:      r.shuttingDown,
+	}
+	// tweak the event correlator, so we don't loose important events.
+	broadcaster := record.NewBroadcasterWithCorrelatorOptions(record.CorrelatorOptions{
```

Comment:
you forgot to modify the key func.  without that you'll still be combining I think

Reply:
i think we want to combine? so the default key function seems fine? or we never want to combine events with same reason?

Reply:
> i think we want to combine? so the default key function seems fine? or we never want to combine events with same reason?

we never want to combine events with different messages.